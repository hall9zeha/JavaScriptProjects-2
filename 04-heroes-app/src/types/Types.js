
//Centralizamos los nombres de las acciones que dispararemos en otros eventos
//Algo parecido  a las constantes Strings que se usan comunmente para manejar desde un solo lugar
//estas constantes que podremos usar en cualquier lugar de la aplicaci√≥n.
export const Types = {
    login:'[auth] Login',
    logout:'[auth] Logout'
}