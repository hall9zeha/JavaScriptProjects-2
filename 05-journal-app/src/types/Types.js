export const types={
    login:'[Auth] Login',
    logout:'[Auth] Logout',

    uiSetError:'[UI] Set error',
    uiRemoveError:'[UI] Remove error',

    // Loading types
    uiStartLoading:'[UI] Start loading',
    uiFinishLoading: '[UI] Finish loading'

}