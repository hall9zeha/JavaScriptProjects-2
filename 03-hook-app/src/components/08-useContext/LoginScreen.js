import React, { useContext } from 'react'
import { UserContext } from './UserContext'

export const LoginScreen = () => {
    const {user, setUser} = useContext(UserContext);

    const actionLogin = ()=>{
        const {id,name,email} = user;
        setUser({
            id,
            name,
            email,
            sigIn:true
        })
    }

  return (
    <div>
        <h1>LoginScreen</h1>
        <hr/>
        <button onClick={actionLogin} className='btn btn-primary '>Login</button>
    </div>
  )
}
