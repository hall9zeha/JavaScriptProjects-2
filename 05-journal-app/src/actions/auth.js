// Las acciones no son más que simples funciones que representan nuestros casos de uso

import { firebase, googleAuthProvider } from "../firebase/firebaseConfig";
import { types } from "../types/Types";
import { uiFinishLoading, uiStartLoading } from "./ui";



export const startLoginWithEmailPassword = (email, password) =>{
    return (dispatch)=>{
        dispatch(uiStartLoading())
        firebase.auth().signInWithEmailAndPassword(email,password)
            .then(({user})=>{
                dispatch(login(user.uid,user.displayName));
                dispatch(uiFinishLoading())
            })
            .catch(e=>{
                dispatch(uiFinishLoading())
                console.log(e)
            })
        
    }
}

export const startLoginWithGoogle = ()=>{
    return (dispatch)=>{
        firebase.auth().signInWithPopup(googleAuthProvider)
            .then(({user})=>{
                dispatch(login(user.uid,user.displayName))
            })
    }
}

export const startRegisterWithEmailPasswordName = (email, password,name)=>{
    return(dispatch)=>{
        firebase.auth().createUserWithEmailAndPassword(email,password)
            .then(async({user}) =>{
                await user.updateProfile({displayName:name});
                dispatch(login(user.uid,user.displayName));
            })
            .catch(e=>{
                console.log(e)
            })
    }
}


// La siguiente función retornará un objeto
export const login = (uid, displayName)=>({
    type: types.login,
    payload:{
        uid,
        displayName
    }
})

// Logout

export const startLogout = ()=>{
    return async (dispatch)=>{
        await firebase.auth().signOut();
        dispatch(logout())
    }
}

export const logout=()=>({
    type:types.logout
})