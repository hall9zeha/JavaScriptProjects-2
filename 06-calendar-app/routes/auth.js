/*
User routes /auth
host + /api/auth

*/

const {Router} = require('express');
const router = Router();
const {check} = require('express-validator')

const {createUser, login,renewToken} = require('../controllers/auth');
const { fieldValidate } = require('../middlewares/field-validators');
const { validateJWT } = require('../middlewares/jws-validator');

router.post('/new',
    //Midlewares para validar los campos necesarios, comenzamos a hacerlos aqu√≠ y los recibimos en nuestros controllers

    [
        check('name', 'Name is required').not().isEmpty(),
        check('email', 'Email is required').isEmail(),
        check('password', 'The password must have 6 or more characters').isLength({min:6}),
        fieldValidate
    ],    
    createUser)
router.post('/',
    [
        check('email', 'Email is required').isEmail(),
        check('password', 'The password must have 6 or more characters').isLength({min:6}),
        fieldValidate
    ],
    login)
    // Validamos el json web token con vlidateJWT 
router.get('/renew',validateJWT, renewToken)

module.exports = router;