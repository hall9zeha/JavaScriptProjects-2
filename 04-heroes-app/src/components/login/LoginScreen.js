import React, { useContext } from 'react'
import { useNavigate } from 'react-router-dom'
import { AuthContext } from '../../auth/authContext';
import { Types } from '../../types/Types';

export const LoginScreen = () => {
    const navigate = useNavigate();

    const {dispatch} = useContext(AuthContext);

    const handleLogin=()=>{
      const action = {
        type:Types.login,
        payload:{name:'Martha'}
      }
      dispatch(action);
      navigate('/marvel',{
          replace:true
      })
    }

  return (
    <div className='container mt-5'>
        <h1>Login</h1>
        <hr/>

        <button className='btn btn-primary ' onClick={handleLogin}>Login</button>
    </div>
  )
}
