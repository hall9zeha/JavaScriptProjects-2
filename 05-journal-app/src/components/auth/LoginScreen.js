import React from 'react'
import { Link } from 'react-router-dom'
import { useForm } from '../../hooks/useForm'
import { useDispatch } from 'react-redux'
import { login, startLoginAsyncExample, startLoginWithGoogle } from '../../actions/auth'

export const LoginScreen = () => {

  const dispatch = useDispatch()


  const [formValues, handleInputChange ] = useForm({
    email:'martha@gmail.com',
    password:12345678
  })

  const {email,password} = formValues

  const handleEventLogin = (e) =>{
    e.preventDefault();
    dispatch(login(123456,'Martha'))
    
    // Ejemplo de acción asíncrona para probar que la configuración middleware en redux funciona correctamente
    // dispatch(startLoginAsyncExample(email,password));
  }
  const handleLoginWithGoogle =()=>{
    dispatch(startLoginWithGoogle());
  }
  return (
    <>
      <h3 className='auth__title'>Login</h3>
      <form onSubmit={handleEventLogin}>
        <input
          type='text'
          placeholder='Email'
          name="email"
          autoComplete='off'
          className='auth__input'
          value={email}
          onChange={handleInputChange}
        />
        <input
          type='password'
          placeholder='Password'
          name='password'
          className='auth__input'
          value={password}
          onChange={handleInputChange}
        />
        <button

          type='submit'
          className='btn btn-primary btn-block'
        >
          Login
        </button>
    
        <div className='auth__social-networks'>
          <p>Login with social networks</p>
              <div  className="google-btn" 
                    onClick={handleLoginWithGoogle}
              >
                  <div className="google-icon-wrapper">
                      <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/archive/c/c1/20210618182605%21Google_%22G%22_logo.svg" alt="google button" />
                  </div>
                  <p className="btn-text">
                      <b>Sign in with google</b>
                  </p>
              </div>
        </div>
        <Link
          to="/auth/register"
          className="link"
        >
          Create new account
        </Link>   
      </form>
    </>
  )
}
