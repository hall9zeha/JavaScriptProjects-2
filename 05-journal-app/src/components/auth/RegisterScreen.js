import React from 'react'
import { Link } from 'react-router-dom'
import { useForm } from '../../hooks/useForm'
import validator from 'validator'
import { useDispatch, useSelector } from 'react-redux'
import { removeError, setError } from '../../actions/ui'
import { startRegisterWithEmailPasswordName } from '../../actions/auth'

export const RegisterScreen = () => {
  const dispatch = useDispatch()
  // Para obtener los objetos dentro de los estados de redux lo hacemos con useSelector de redux
  const {msgError} = useSelector(state=>state.ui);
  

  const [values, handleInputChange] = useForm({
    name:'Martha',
    email:'martha@mail.com',
    password:'123456',
    confirm:'123456'
  })
  const {name,email,password,confirm} = values
  const handleRegister=(e)=>{
    e.preventDefault();
    if(isFormValid()){
      dispatch(startRegisterWithEmailPasswordName(email,password,name))
    }
    
  }
  const isFormValid = () =>{
    if(name.trim().length ===0){
      dispatch(setError("Name is required"))
      return false;
    } else if(!validator.isEmail(email)){
      dispatch(setError("Email is not valid"))
      return false;
    }else if(password !== confirm || password.length <5){
      dispatch(setError("Password should be at least of 6 characters and match each other"))
      return false;
    }
    dispatch(removeError())
    return true;

  }

  return (
    <>
    <h3 className='auth__title'>Register</h3>
    <form onSubmit={handleRegister}>
      {/* Si el msgError no es nulo entonces mostrar√° un elemento div con el mesaje */}
      {
        msgError && (
          <div className='auth__alert-error'>{msgError}</div>
        )
      }
    <input
        type='text'
        placeholder='Name'
        name="name"
        autoComplete='off'
        className='auth__input'
        value={name}
        onChange={handleInputChange}
      />
      <input
        type='text'
        placeholder='Email'
        name="email"
        autoComplete='off'
        className='auth__input'
        value={email}
        onChange={handleInputChange}
      />
      <input
        type='password'
        placeholder='Password'
        name='password'
        className='auth__input'
        value={password}
        onChange={handleInputChange}
      />
       <input
        type='password'
        placeholder='Confirm password'
        name="confirm"
        autoComplete='off'
        className='auth__input'
        value={confirm}
        onChange={handleInputChange}
      />
      <button
        type='submit'
        className='btn btn-primary btn-block mb-5'
      >
        Register
      </button>
  
      <Link
        to="/auth/login"
        className="link"
      >
        Already registered?
      </Link>   
    </form>
  </>
  )
}
