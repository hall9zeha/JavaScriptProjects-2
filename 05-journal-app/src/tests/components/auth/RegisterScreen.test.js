/**
 * @jest-environment jsdom
 */
import {Provider} from 'react-redux';
import renderer ,{act}from 'react-test-renderer';
import { LoginScreen } from '../../../components/auth/LoginScreen';
import {MemoryRouter} from 'react-router-dom';

import configureStore from 'redux-mock-store'
import { render, fireEvent } from "@testing-library/react";
import { RegisterScreen } from '../../../components/auth/RegisterScreen';
import { types } from '../../../types/Types';



const thunk = require('redux-thunk').thunk;
//***************************/
const middlewares = [thunk]
const mockStore = configureStore(middlewares)

const initState = {
    auth:{},
    ui:{
        loading:false,
        msgError:null
    }
}
let store = mockStore(initState)
//Para simular el dispatch y evaluar como son llamadas las funciones 
//En esta prueba las acciones llamadas son síncronas por eso no la usaremos
//store.dispatch=jest.fn()

jest.mock('../../../actions/auth',()=>({
    startLoginWithGoogle:jest.fn(),
    startLoginWithEmailPassword:jest.fn()
}))


describe('Test in <RegisterScreen/> component', () => { 
    beforeEach(()=>{
         
            jest.clearAllMocks();
        }
    )
    const wrapper = renderer.create( <Provider store={store}>
        <MemoryRouter>
            <RegisterScreen/>
        </MemoryRouter>
    </Provider>)

    test('should show successfully', () => { 
        expect(wrapper).toMatchSnapshot()
     })
    test('should dispatch of respective action', async() => { 
        const txtName = await wrapper.root.findAllByProps({name:'name'});//Devuelve un array debemos escoger la primera posición [0]
        const txtEmail = await wrapper.root.findAllByProps({name:'email'})//Devuelve un array debemos escoger la primera posición [0]
        const form = await wrapper.root.findByType('form')
        const value='Barry'
        
        act(()=>{
            txtName[0].props.onChange({target:{
                value:value,
                name:'name'
            }})
        })
        act(()=>{
            txtEmail[0].props.onChange({target:{
                value:'',
                name:'email'
            }})
        })
        form.props.onSubmit({preventDefault:jest.fn()})
        const actions = store.getActions();
        console.log(actions)
        expect(actions[0]).toEqual({
            type:types.uiSetError,
            payload:'Email is not valid'
        })
        
     })
     test('should show alert box with error', async() => {

            const initState = {
                auth:{},
                ui:{
                    loading:false,
                    msgError:'Email no es válido'
                }
            }
            const store = mockStore(initState)
            const wrapper = renderer.create( <Provider store={store}>
                <MemoryRouter>
                    <RegisterScreen/>
                </MemoryRouter>
            </Provider>)
            
            const errorBox =  await wrapper.root.findByProps({className:'auth__alert-error'})
            
            expect(errorBox).toBeTruthy()
            expect(errorBox.props.children.trim()).toBe(initState.ui.msgError)

      })
 })