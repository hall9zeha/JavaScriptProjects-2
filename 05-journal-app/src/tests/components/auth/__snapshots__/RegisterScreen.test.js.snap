// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in <RegisterScreen/> component should show alert box with error 1`] = `
Array [
  <h3
    className="auth__title"
  >
    Register
  </h3>,
  <form
    className="animate__animated animate__fadeIn animated__faster"
    onSubmit={[Function]}
  >
    <div
      className="auth__alert-error"
    >
      Email no es v√°lido
    </div>
    <input
      autoComplete="off"
      className="auth__input"
      name="name"
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value="Martha"
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="text"
      value="martha@mail.com"
    />
    <input
      className="auth__input"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value="123456"
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="confirm"
      onChange={[Function]}
      placeholder="Confirm password"
      type="password"
      value="123456"
    />
    <button
      className="btn btn-primary btn-block mb-5"
      type="submit"
    >
      Register
    </button>
    <a
      className="link"
      href="/auth/login"
      onClick={[Function]}
    >
      Already registered?
    </a>
  </form>,
]
`;

exports[`Test in <RegisterScreen/> component should show successfully 1`] = `
Array [
  <h3
    className="auth__title"
  >
    Register
  </h3>,
  <form
    className="animate__animated animate__fadeIn animated__faster"
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      name="name"
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value="Martha"
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="text"
      value="martha@mail.com"
    />
    <input
      className="auth__input"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value="123456"
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="confirm"
      onChange={[Function]}
      placeholder="Confirm password"
      type="password"
      value="123456"
    />
    <button
      className="btn btn-primary btn-block mb-5"
      type="submit"
    >
      Register
    </button>
    <a
      className="link"
      href="/auth/login"
      onClick={[Function]}
    >
      Already registered?
    </a>
  </form>,
]
`;
